ciongke/cmd/test-pr/main.go

```diff
@@ -0,0 +1,395 @@
+	"gopkg.in/yaml.v2"
+
+	"github.com/kubernetes/test-infra/ciongke/cmdlog"
+	"github.com/kubernetes/test-infra/ciongke/gcs"
+	"github.com/kubernetes/test-infra/ciongke/github"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+var (
+	repoURL       = flag.String("repo-url", "", "URL of the repo to test.")
+	repoOwner     = flag.String("repo-owner", "", "Owner of the repo to test.")
+	repoName      = flag.String("repo-name", "", "Name of the repo to test.")
+	pr            = flag.Int("pr", 0, "Pull request to test.")
+	branch        = flag.String("branch", "", "Target branch.")
+	head          = flag.String("head", "", "Head SHA.")
+	resultsBucket = flag.String("results-bucket", "", "Bucket to store job results.")
+	sourceBucket  = flag.String("source-bucket", "", "Bucket to store PR source.")
+	namespace     = flag.String("namespace", "", "Namespace for created pods.")
+	workspace     = flag.String("workspace", "/workspace", "Where to checkout the repo.")
+	dryRun        = flag.Bool("dry-run", true, "Whether or not to avoid mutating calls to GitHub.")
+
+	githubTokenFile = flag.String("github-token-file", "/etc/oauth/oauth", "Path to the file containing the GitHub OAuth secret.")
+)
+
+const (
+	watchTimeout = 10 * time.Hour
+)
+
+type jobDescription struct {
+	Name        string `yaml:"name"`
+	Description string `yaml:"description"`
+	Image       string `yaml:"image"`
+}
+
+type testClient struct {
+	GitHubClient    github.Client
+	KubeClient      client.Interface
+	KubeBatchClient client.BatchInterface
+}
+
+func main() {
+	flag.Parse()
+
+	oauthSecretRaw, err := ioutil.ReadFile(*githubTokenFile)
+	if err != nil {
+		log.Fatalf("Could not read oauth secret file: %s", err)
+	}
+	oauthSecret := string(bytes.TrimSpace(oauthSecretRaw))
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: oauthSecret})
+	tc := oauth2.NewClient(oauth2.NoContext, ts)
+	var githubClient github.Client
+	if *dryRun {
+		githubClient = github.NewDryRunClient(tc)
+	} else {
+		githubClient = github.NewClient(tc)
+	}
+
+	clonePath := filepath.Join(*workspace, "repo")
+	mergeable, err := checkoutPR(*repoURL, clonePath, *branch, *head, *pr)
+	if err != nil {
+		log.Printf("Error preparing git repo: %s", err)
+		return
+	}
+	if !mergeable {
+		log.Printf("PR %d needs rebase, skipping tests.", *pr)
+		return
+	}
+
+	sourcePath := filepath.Join(*workspace, *head+".tar.gz")
+	tar := exec.Command("tar", "czf", sourcePath, "repo")
+	tar.Dir = *workspace
+	if err = cmdlog.RunWithLogs(tar); err != nil {
+		log.Printf("Failed to create source archive: %s", err)
+		return
+	}
+	if err = gcs.UploadSource(*sourceBucket, *head, sourcePath); err != nil {
+		log.Printf("Failed to upload source: %s", err)
+		return
+	}
+
+	b, err := ioutil.ReadFile(filepath.Join(clonePath, ".test.yml"))
```

Comment:
nit: Why do you want this file to be hidden? Wouldn't `test.yml` be easier for people to see/understand?